### PR TITLE
Allow return in tailrec position

### PR DESCRIPTION
**Pull request by mbovel**
Implement https://github.com/lampepfl/dotty-feature-requests/issues/261.

Joint work with @ghostbuster91 in the framework of the issue spree.

### CODE REVIEW

File: compiler/src/dotty/tools/dotc/transform/TailRec.scala [hunk 1]
```diff
@@ -444,7 +444,7 @@ class TailRec extends MiniPhase {
 
         case Return(expr, from) =>
           val fromSym = from.symbol
-          val inTailPosition = fromSym.is(Label) && tailPositionLabeledSyms.contains(fromSym)
+          val inTailPosition = true
```
Comment (sjrd) [inline compiler/src/dotty/tools/dotc/transform/TailRec.scala:447]: This change will consider as tail-call a `return@label` where the label is not itself in tail position. This is incorrect. It should be something like
```suggestion
          val inTailPosition = !fromSym.is(Label) || tailPositionLabeledSyms.contains(fromSym)
```

**Review by sjrd — CHANGES_REQUESTED** [review]

File: compiler/src/dotty/tools/dotc/transform/TailRec.scala (hunk 1 above)
Comment (mbovel) [reply compiler/src/dotty/tools/dotc/transform/TailRec.scala]: Done ✅

**Review by sjrd — COMMENTED** [review]
This shouldn't create any issues now.

However, to be complete, you'll have to actually recurse in all trees all the time. See this comment:
https://github.com/lampepfl/dotty/blob/a34977551e9e13090e6b5201f7d91879464f991c/compiler/src/dotty/tools/dotc/transform/TailRec.scala#L29-L31
Now any nested tree could contain a recursive call, which was not the case before.

Also, this will need more tests before it's ready to merge. ;)

**Review by sjrd — APPROVED** [review]
Looks good to me.

There might be concerns about the performance impact, since now checking needs to be done everywhere all the time. So perhaps ask the performance bot?

**Comment by som-snytt** [comment]
This is a great first step in the rehabilitation of `return`. sjrd's eyes make me confident that the feature is not just catering to folks accustomed to the keyword. The other `return` issue is to allow `return` followed by miscellaneous defs. https://github.com/lampepfl/dotty-feature-requests/issues/256 I will pursue that, inspired by this PR.

The t-shirt should make reference to Nietzsche, "The eternal `return`". I don't know if PRs have gotten their own custom t-shirt until this point.

**Comment by mbovel** [comment]
test performance please

**Comment by mbovel** [comment]
@sjrd could we merge this or do we need more changes or measurements?

**Review by sjrd — APPROVED** [review]
The benchmark results look good to me. It's good to go as far as I'm concerned.

**Pull request merged**